### DISCUSSION TITLE
Cachy kernels for NixOS

### DISCUSSION THREAD

**Discussion by drakon64** [CachyOS/linux-cachyos#273] [Q&A] [answered]
I'm currently attempting to package the Cachy kernels (default and LTS) for NixOS (https://github.com/NixOS/nixpkgs/pull/330917). While these compile and run (haven't tested LTS yet) fine, a point was raised about how some of the optimisations from the Cachy kernel might be coming from the kernel config tweaks that come from upstream Arch and wouldn't be included in the NixOS version as a result.

Currently the kernel config we're applying matches the default applied by the PKGCONFIG script (far as I can tell). Is there anything we're possibly missing by exclusively following the PKGCONFIG?

We're also enabling the following features as modules for the default kernel:
- `I2C_NCT6775`
- `NTSYNC`
- `V4L2_LOOPBACK`

and for the LTS kernel:
- `I2C_NCT6775`
- `STEAMDECK`

Are there any additional modules that Cachy adds which we've missed?

**Comment by ptr1337** [comment]
Hi @drakon64 ,

Thanks for working on providing the CachyOS Kernel to nixpkgs.
There are some different configs enabled, this is _mostly_ done via the PKGBUILD and scripts/config.
Following get enabled via the PKGBUILD:

1. `CACHY`
2. `CC_OPTIMIZE_FOR_PERFORMANCE_O3`
3. `TCP_CONG_BBR DEFAULT_BBR, --set-str DEFAULT_TCP_CONG bbr`
4. `TRANSPARENT_HUGEPAGE_ALWAYS`
5. NTSYNC
6. V4L2_LOOPBACK
7. I2C_NCT6775
8. SteamDeck related config variables

for sched-ext and BORE support, which is as default used in "linux-cachyos" kernel:
`-e SCHED_CLASS_EXT -e SCHED_BORE --set-val MIN_BASE_SLICE_NS 1000000`

Be aware, that for sched-ext support there is also BPF and BTF support required.

For T2 Support see following:
https://github.com/t2linux/linux-t2-patches/blob/main/extra_config

**Comment by drakon64** [comment]
Thank you for the response @ptr1337.

I have a few questions.

~~In https://github.com/CachyOS/linux-cachyos/blob/3728444ae7d3ba1ac6b6fa7401089143597b5aba/linux-cachyos/PKGBUILD, I don't see where NTSYNC is enabled. Is this enabled by default when the patches are applied?~~ Found the relevant config in the patch files.

Also, I don't see the Steam Deck module in https://github.com/CachyOS/linux-cachyos/blob/3728444ae7d3ba1ac6b6fa7401089143597b5aba/linux-cachyos/PKGBUILD but I do see it in https://github.com/CachyOS/linux-cachyos/blob/3728444ae7d3ba1ac6b6fa7401089143597b5aba/linux-cachyos-lts/PKGBUILD. I was under the impression that the Steam Deck patches were mainlined now, is it just that the 6.6 kernel is from before the patches were mainlined and the 6.10 kernel has them anyway?

There's also two other points I wanted to bring up:
1. Per https://github.com/NixOS/nixpkgs/pull/330917#pullrequestreview-2207460947, our kernel configuration might be problematic when diagnosing issues. We use the 'standard' NixOS kernel config and apply CachyOS's PKGBUILD config on top of that rather than using the base Arch config. Is this going to cause an issue?
2. I realise there's no CachyOS AArch64, but would these patches be expected to work on AArch64 and would this be supported?

**Reply by ptr1337** [reply, marked as answer]
Hey!

Yes, currently this is not fully consistent, since we only made user choices "changeable" via the PKGBUILD and set over there.
NTSync is enabled in our config directly:
https://github.com/CachyOS/linux-cachyos/blob/master/linux-cachyos/config#L2775

The SteamDeck Module is as default enabled, since it got merged upstream, even tough there are more and newer patches but these are handled in an own kernel to not bloat the patchset.
You can find it here: https://github.com/CachyOS/kernel-patches/blob/master/6.10/misc/0001-handheld.patch
I do not suggest to apply this patch as default.

1. Generally, I would say this is fine. "Chatoic-Nyx" already have a linux-cachyos flake, but they are following our config and apply NixOS related changes on their own to the config. As long there are not majority changes, this is fine from my side. Every distro has its own kernel config and this is fine.
2. We did so far not any compilation tests against aarch64 yet. This is generally planned in the future to enhance it. As long it compiles on your side - go for it.

I would also suggest you to apply the BORE Scheduler and sched-ext Framework, which is applied on our default kernel too.

**Reply by drakon64** [reply]
Thanks for the help @ptr1337!
